### PR TITLE
2.2.0

### PR DESCRIPTION
**Pull request by AuthorOfTheSurf**
- [x] Separate out compiler from command line interface #57  
- [x] Allow `string` and `numeric` literals to be keys to dictionaries #59 (issue) #61 (PR)
- [x] Review `Types` module and literals
- [x] Update `is` operator #54 (issue) #64 (PR) 
- [x] Parsing bugs fixed #64 

_Postponing cli improvements till 2.3_
- ~~\- Improve cli with additional options~~
  - ~~\- Output to file~~
  - ~~\- Directory to directory, e.g. `/ks` to `/js`~~

_Aiming to push release on 2015-07-31 for code-output improvements_

### CODE REVIEW

**Comment by jadengore** [comment]
Are you looking to merge this soon?

**Comment by AuthorOfTheSurf** [comment]
I want to get in the changes to the `is` operator. Would like to merge it as soon as that is done.

File: entities/object-literal.js [hunk 1]
```diff
@@ -18,23 +19,18 @@ ObjectLiteral.prototype.analyze = function (context) {
...
-    var initializer = p.initializer.generateJavaScript(state)
-
-    propText.push(name + ':' + initializer)
+    props.push(p.generateJavaScript(state))
   })
 
-  if (propText) {
-    js.push(propText.join(', '))
+  if (props) {
+    js.push(props.join(', '))
```
Comment (jadengore) [inline entities/object-literal.js]: Some great cleanup here :+1:

File: entities/type.js [hunk 2]
```diff
@@ -14,61 +14,11 @@ Type.prototype.toString = function () {
...
-  // In more sophisticated languages, comapatibility would be more complex
-  return this == otherType;  
-}
-
-Type.prototype.mustBeCompatibleWith = function (otherType, message, location) {
-  if (!this.isCompatibleWith(otherType)) {
-    error(message, location)
-  }
-}
+exports.CLOSURE = Type.CLOSURE = new Type('CLOSURE')
```
Comment (jadengore) [inline entities/type.js]: McGarbage to Cleanliness :100:

File: parser.js [hunk 3]
```diff
@@ -61,10 +58,10 @@ function parseProgram() {
     error('KobraScript programs may not be empty')
   }
 
-  var initialBlock = new Block(parseStatements())
+  var programBlock = new Block(parseStatements())
   match('EOF')
```
Comment (jadengore) [inline parser.js]: Yes! This is more explicit for what it is: the block that represents the highest level, which is the program

File: scanner.js [hunk 4]
```diff
@@ -185,15 +190,26 @@ function scan(line, linenumber, tokens) {
...
+    }
+    else if (atNull) {
+      var lexeme = 'null'
+      emit('NULLLIT', lexeme)
+      pos += lexeme.length
+    }
+    else if (atUndefined) {
+      var lexeme = 'undefined'
+      emit('UNDEFINEDLIT', lexeme)
+      pos += lexeme.length
```
Comment (jadengore) [inline scanner.js]: So does this mean null and undefined are being tokenized differently now?

File: test/kobra-code/good-programs/is-operator-negated.ks [hunk 5]
```diff
@@ -0,0 +1,5 @@
+$ x = 10
+
+loge x is 'number' && x is 'number'
+loge !(x is 'number') && !x 
+loge !!(x is 'number')
```
Comment (jadengore) [inline test/kobra-code/good-programs/is-operator-negated.ks]: Wow, a lot of love for `loge`, get some `say` in there~

File: test/kobra-code/good-programs/negated-object.ks [hunk 6]
```diff
@@ -0,0 +1,2 @@
+$ x = !{}
```
Comment (jadengore) [inline test/kobra-code/good-programs/negated-object.ks]: What an interesting test!

File: test/kobra-code/good-programs/very-negated.ks [hunk 7]
```diff
@@ -0,0 +1,4 @@
+$ zen = true
+
+!!!zen
+!!!!!!!!zen
```
Comment (jadengore) [inline test/kobra-code/good-programs/very-negated.ks]: `very-negated` is a truuuuuu filename

**Comment by jadengore** [comment]
I think besides my note [here](https://github.com/AuthorOfTheSurf/KobraScript/pull/58#discussion-diff-36161605), I think this puppy is ready to ship. Really great cleanup and refactoring. KobraScript is really starting to look a lot cleaner... now just some CLI work and we will be golden :shipit:

File: scanner.js (hunk 4 above)
Comment (AuthorOfTheSurf) [inline scanner.js]: Yes, previously they did not have their type as the `kind`, see here in the parser where [we used to look for the lexeme, not the type](https://github.com/AuthorOfTheSurf/KobraScript/pull/58/files#diff-09461573a85e2d94f056dd6814769042R519) as in the surrounding code.

visual:
`{ kind: "UNDEFINEDLIT", lexeme: "undefined" }`
`{ kind: "undefined", lexeme: "undefined" }`

File: test/kobra-code/good-programs/very-negated.ks (hunk 7 above)
Comment (AuthorOfTheSurf) [inline test/kobra-code/good-programs/very-negated.ks]: :moneybag: :100:

**Comment by AuthorOfTheSurf** [comment]
Roger that, looking forward to the cli improvements. Now if only we could get someone to write our syntax highlighting...

File: test/kobra-code/good-programs/is-operator-negated.ks (hunk 5 above)
Comment (AuthorOfTheSurf) [inline test/kobra-code/good-programs/is-operator-negated.ks]: Maybe in 2.3!

**Comment by ebramanti** [comment]
@AuthorOfTheSurf [did you see my repo?](https://github.com/jadengore/language-kobrascript)

**Pull request merged**